File: MATLAB/automorph.m
% This is the main Automorph file to extract morphometrics from Lidar
% profil   es.
%
% Michael Itzkin, 3/26/2018
%------------------------------------------------------------------------%
tic;
clear all
close all
clc

% First load all the excel files to be used. Store the Excel files in the 
% "Excel files" folder. Automorph will make any other folder as it goes
addpath('Subfunctions')
addpath('Excel Files')
files = dir(['Excel Files' filesep '*.xlsx']);

% Set the location of the state plane coordinates. Check the "sp_proj"
% function in the Subfunctions folder for details
sp_loc = 'north carolina'; %% You can change this!!

% Set the backshore drop and buffer_pct for the crest_finder
backshore_drop = 0.3; %% You can change this!! (0.6 in Mull/Ruggiero 2014)
buffer_pct = 0.10; %% Can change this!!

% Set the dune slope threshold to ignore buildings (Stockdon et al, 2009)
dune_slope_threshold = 35;  % Degrees (35 in Stockdon et al, 2009)

% Set the starting year of analysis and the final year
start_year = '1997';
final_year = '2016';

% Set the number of columns for the morphometrics table, this should match
% the number of points being saved in the morpho_table_maker script.
% Shouldn't matter if the number is incorrect but the program will run
% faster if it is
num_cols = 59; %% You can change this!!

% Loop through the Excel files, ignoring the Profile Info Table
for q = 1:length(files)
    if ~strcmp(files(q).name, 'Profile Info Table.xlsx') &&...
            ~strcmp(files(q).name(end-10:end-5), 'Fences')
        
        % Excelload will open the .xlsx file with the profiles and format
        % it into a 3D matrix with a dimension for X, Y, and Z. If this has
        % been done before on the given .xlsx file than it will just load
        % the formatted data and move on.
        excelload
        
        % Load the morpho_table from the previous year, skip the header row.
        if ~strcmp(year, start_year)
           if strcmp(year, '1998')
               previous_year = '1997';
           elseif strcmp(year, '1999')
               previous_year = '1998';
           elseif strcmp(year, '2000')
               previous_year = '1999';
           elseif strcmp(year, '2004')
               previous_year = '2000';
           elseif strcmp(year, '2005')
               previous_year = '2004';
           elseif strcmp(year, '2010')
               previous_year = '2005';
           elseif strcmp(year, '2011')
               previous_year = '2010';
           elseif strcmp(year, '2014')
               previous_year = '2011';
           elseif strcmp(year, '2016')
               previous_year = '2014';
           end
           original_data_fname = sprintf('%s%s%s%sMorphometrics for %s %s.csv',...
               location(1:end-1), filesep, previous_year, filesep,...
               location(1:end-1), previous_year);
           original_data = csvread(original_data_fname, 1);
        end
        
        % Preallocate an empty matrix to loop data into
        morpho_table = NaN(transects, num_cols);
        
        % If the area has fences, find where all the fences are in the
        % study area. If this script ran before for the current study area
        % it will just load the "Fence Crossings" .mat file for the
        % study area
        if fenced
            fence_finder
        end
        clc
        
        local_x_values = NaN(max_index-2, transects);

        for pass = 1:2 %% Change to a 2 to use adjustment codes
            % Loop through the transects and plot the morphometrics
            
            % Initialize the progress bar
            progress_bar_label = sprintf('%s %s',...
                    location(1:end-1), year);
            progressbar(progress_bar_label)
            
            for k = 1:length(profiles(1,:,:))
                
                % Update the progress bar
                progressbar(k/length(profiles(1,:,:)))
                
                % Smooth the profile with a moving filter 
                profiles(:,k,3) = smooth(profiles(:,k,3),11,'moving');

                % Find MHW
                mhw_finder
                
                % Find the natural dune crest
                crest_finder
                
                % Find the natural dune toe
                toe_finder
                
                % Check that the dune slope is less than
                % the threshold value. Keep finding new crests
                % until it is
                dune_slope = atand((y_crest - y_toe) / abs(x_crest - x_toe));
                while (dune_slope >= dune_slope_threshold) &&...
                        (ignore_slope_threshold == 0)
                    
                    % Find the natural dune crest
                    crest_finder_slope_threshold
                    
                    % Find the natural dune toe
                    toe_finder
                    
                    % Recalculate the dune slope
                    dune_slope = atand((y_crest - y_toe) / abs(x_crest - x_toe));
                end

                % Find an possible berm
                berm_finder
                
                % Find the natural dune heel
                heel_finder

                % Find the natural dune volume 
                natural_dune_volume_finder               
                
                % If a fence exists and is seaward of
                % the natural dune crest, run extra methods
                if fenced 
                    fence_locator % Find where the fence is
                    if exist('fence_index') && (fence_index >= crest_index)
                        fence_crest_finder % Find where the fenced dune crest is
                        fence_heel_finder % Find the fenced dune heel
                        fence_toe_finder % Find the fenced dune toe
                        fenced_dune_volume_finder % Find the fenced dune volume
                        total_dune_volume_finder % Find the fenced+natural volume
                    end
                end

                % Calculate the beach width
                beach_width_finder
                
                % Calculate the beach and foreshore slope
                beach_slope_finder
                
                % Find the elevation of the crest height from the starting year
                find_starting_crest_location
                
                % Put the data into the morpho_table
                morpho_table_maker

                % Plot the profile with the morphometrics
                profile_plotter

                % Clear extraneous variables
                clear fence_heel_index x_fence_heel y_fence_heel...
                    local_x_fence_heel local_y_fence_heel fence_heel_lon...
                    fence_heel_lat fence_toe_index x_fence_toe y_fence_toe...
                    local_x_fence_toe local_y_fence_toe fence_toe_lon...
                    fence_toe_lat y_fence_crest fence_crest_index

            end
        end
        
        % Save the morphometrics to a .csv and .mat file
        save_morphometrics
        
        % Plot a 2D view of the results
        morpho_2d_plotter
        
        % Plot the crest_threshold over time
        crest_threshold_plotter
        
        % Clear profiles for the next location/year and the local_x_values
        clear profiles local_x_values
        
        % Move the file out of Excel files and into the location folder
        source = sprintf('Excel Files%s%s', filesep, files(q).name);
        destination = sprintf('%s%s%s', location(1:end-1), filesep, year);
        movefile(source, destination, 'f');
        
        % If it is the final year and there are fences, move the fences
        % file to the location folder as well
        fence_file = sprintf('Excel Files%s%s Fences.xlsx',...
            filesep, location(1:end-1));
        if strcmp(year, final_year) && exist(fence_file)
            movefile(fence_file, location(1:end-1),'f');
        end
        
    end   
end

toc;
sec = rem(toc,60);
min = floor(toc/60);
hr = floor(min/60);
min = min - (hr*60);
fprintf('The program completed in %d hour(s), %d minute(s), and %.2f seconds\n',...
    hr, min, round(sec))

File: Python/Automorph.py
"""
Automatically extract dune and beach morphometrics
from LiDAR profiles

Michael Itzkin, 6/28/2021
"""

from Functions import Data_Functions as dfuncs
from Functions import Morpho_Functions as mfuncs
from Functions import Plot_Functions as pfuncs

import pandas as pd
import shutil
import os


# Set paths to other folders
DATA_DIR = os.path.join('..', 'Data')


def main():
    """
    Run the analysis
    """

    # Set parameters
    use_extension = '.xyz'
    epsg = 3358             # St. Pete = 2778, NC = 3358
    grid_size = 0.5         # Doesn't matter: The gridded profiles aren't used
    mhw = 0.34              # St. Pete = 0.187, NC = 0.34
    heel_threshold = 0.6
    crest_pct = 0.1

    # Loop through the files in the data directory. Only
    # consider profiles with an .xyz extension
    for file in os.listdir(DATA_DIR):
        if file.endswith(use_extension):

            # Setup a dictionary with blank lists to loop into
            morpho = dfuncs.morpho_dict()

            # Get basic information about the current set of profiles and
            # make a folder to store results in
            location, year, num_profiles = dfuncs.get_basic_information(file)

            # Print out a header to the terminal
            print('\n------------------------------------------------')
            print(f'Currently Working On: {location} {year}')
            print(f'Profiles: {num_profiles}')
            print('------------------------------------------------')

            # Parse out the individual profiles from the main file
            # into individual .txt files
            dfuncs.make_profile_files(file, location, year, num_profiles, epsg)

            # Loop over the profiles
            for profile in range(1, num_profiles + 1):

                # Store the profile number
                morpho['Profile'].append(profile)

                # Determine the profile length and interpolate onto a grid
                # of a pre-defined spacing
                dist_cross, elev_cross, ex, why, lats, lons =\
                    dfuncs.setup_profile(location, year, profile, grid_size)

                # Identify the MHW contour. This function also calculates
                # the foreshore slope since the error method for MHW includes
                # calculating it.
                morpho = mfuncs.find_mhw(morpho, dist_cross, elev_cross,
                                         lats, lons, mhw)

                # Identify the dune crest
                morpho = mfuncs.find_crest(morpho, dist_cross, elev_cross,
                                           lats, lons, mhw,
                                           heel_threshold, crest_pct)

                # Identify the dune heel
                morpho = mfuncs.find_heel(morpho, dist_cross, elev_cross,
                                          lats, lons)

                # Identify the dune toe
                morpho = mfuncs.find_toe(morpho, dist_cross, elev_cross,
                                         lats, lons)

                # Calculate volumes
                morpho = mfuncs.dune_volume(dist_cross, elev_cross, morpho)
                morpho = mfuncs.beach_volume(dist_cross, elev_cross, morpho)
                morpho = mfuncs.profile_volume(dist_cross, elev_cross, morpho)

                # Calculate the profile bearing from heel to MHW
                morpho = mfuncs.orientation(morpho, dist_cross, lats, lons)

                # Plot the profile
                pfuncs.plot_profile(morpho, dist_cross, elev_cross, location,
                                    year, profile, mhw, save=True)

            # Convert morpho to a DataFrame
            df = pd.DataFrame.from_dict(morpho)

            # Calculate metrics that can be done without
            # looping through the profiles
            df['Dune Height'] = df['YCrest'] - df['YToe']
            df['Dune Width'] = df['XToe'] - df['XHeel']
            df['Dune Aspect Ratio'] = df['Dune Height'] / df['Dune Width']
            df['Dune Face Slope'] = df['Dune Height'] / (df['XToe'] - df['XCrest'])
            df['Beach Width'] = df['XMHW'] - df['XToe']
            df['Beach Slope'] = (df['YToe'] - df['YMHW']) / df['Beach Width']

            # Save the DataFrame
            path = os.path.join('..', f'{location}', f'{year}')
            fname = os.path.join(path, f'Morphometrics for {location} {year}.csv')
            df.to_csv(fname, index=False)

        # Move the .txt file to the location and year sub-folder
        src = os.path.join(DATA_DIR, file)
        dst = os.path.join('..', f'{location}', f'{year}', file)
        shutil.move(src, dst)


if __name__ == '__main__':
    main()
